hs-bindgen/src-internal/HsBindgen/Frontend/Analysis/Typedefs.hs

```diff
@@ -0,0 +1,287 @@
+    -- name of the struct, as normally the typedef is the main type and intended
+    -- to be used throughout the code.
+  | ByValue <- valOrRef, [_] <- useSites
+  = let newName   = typedefName
+        newOrigin = updateOrigin taggedName taggedOrigin
+        newTagged = TaggedType{
+                        taggedName   = newName
+                      , taggedOrigin = newOrigin
+                      , taggedKind
+                      }
+    in mempty{
+           squash = Map.singleton typedefName (fromTaggedType newTagged)
+         , rename = Map.singleton taggedName (newName, newOrigin)
+         }
+
+    -- if there are multiple uses, and the names don't match, don't do anything
+  | otherwise
+  = mempty
+
+-- | Update origin information for renamed tagged datatype
+--
+-- If we rename a datatype with a name which was /already/ not original, we
+-- leave the origin information unchanged.
+updateOrigin :: CName -> C.NameOrigin -> C.NameOrigin
+updateOrigin oldName = \case
+    C.NameOriginInSource           -> C.NameOriginRenamedFrom oldName
+    C.NameOriginGenerated   anonId -> C.NameOriginGenerated   anonId
+    C.NameOriginRenamedFrom orig   -> C.NameOriginRenamedFrom orig
+
+{-------------------------------------------------------------------------------
+  Internal auxiliary: tagged types
+
+  TODO: Should we use this elsewhere also?
+-------------------------------------------------------------------------------}
+
+data TaggedType = TaggedType {
+      taggedKind   :: TaggedKind
+    , taggedName   :: CName
+    , taggedOrigin :: C.NameOrigin
+    }
+
+data TaggedKind = Struct | Union | Enum
+
+toTaggedType :: C.Type ResolveBindingSpec -> Maybe TaggedType
+toTaggedType = \case
+    C.TypeStruct name origin -> Just $ TaggedType Struct name origin
+    C.TypeUnion  name origin -> Just $ TaggedType Union  name origin
+    C.TypeEnum   name origin -> Just $ TaggedType Enum   name origin
+    _otherwise               -> Nothing
+
+fromTaggedType :: TaggedType -> C.Type HandleTypedefs
+fromTaggedType TaggedType{..} =
+    case taggedKind of
+      Struct -> C.TypeStruct taggedName taggedOrigin
+      Union  -> C.TypeUnion  taggedName taggedOrigin
+      Enum   -> C.TypeEnum   taggedName taggedOrigin
+
+taggedNameKind :: TaggedKind -> C.NameKind
+taggedNameKind = \case
+    Struct -> C.NameKindStruct
+    Union  -> C.NameKindUnion
+    Enum   -> C.NameKindEnum
+
+origQualId :: TaggedType -> QualDeclId
+origQualId TaggedType{..} =
+    QualDeclId
+      (origDeclId taggedName taggedOrigin)
+      (taggedNameKind taggedKind)
+
+{-------------------------------------------------------------------------------
+  Internal auxiliary: restore original names
+
+  This allows us to look up the type in the 'DeclUseGraph'.
+  TODO: Maybe this should live somewhere more general?
+-------------------------------------------------------------------------------}
+
+origDeclId :: CName -> C.NameOrigin -> DeclId
```

Comment:
@TravisCardwell I suspect you might have a version function somewhere in your `PrelimDeclId` refactoring branch.